crates/loans/src/lib.rs

```diff
@@ -1359,24 +1361,16 @@ impl<T: Config> Pallet<T> {
         Ok(total_borrow_value)
     }
 
-    fn collateral_balance(
-        asset_id: CurrencyId<T>,
-        lend_token_amount: BalanceOf<T>,
-    ) -> Result<BalanceOf<T>, DispatchError> {
-        let exchange_rate = Self::exchange_rate_stored(asset_id)?;
-        let underlying_amount = Self::calc_underlying_amount(lend_token_amount, exchange_rate)?;
-        let market = Self::market(asset_id)?;
-        let effects_amount = market.collateral_factor.mul_ceil(underlying_amount);
-
-        Ok(BalanceOf::<T>::saturated_from(effects_amount))
-    }
-
     fn collateral_amount_value(
-        asset_id: CurrencyId<T>,
-        lend_token_amount: BalanceOf<T>,
+        voucher: &Amount<T>,
     ) -> Result<Amount<T>, DispatchError> {
-        let effects_amount = Self::collateral_balance(asset_id, lend_token_amount)?;
-        Self::get_asset_value(asset_id, effects_amount)
+        let asset_id = Self::underlying_id(voucher.currency())?;
+
+        let underlying = voucher.convert_to(asset_id)?;
+        let market = Self::market(asset_id)?;
+        let effects = underlying.map(|x| market.collateral_factor.mul_ceil(x));
```

Comment:
LGTM overall, this does touch lots of the critical logic and would've been good to get this in before the audit. The multiplication on this line is not hard to miss, for example

Reply:
Yea the audit is my main point of concern. We could ask them if they mind changing commit versions - it hasn't been long since kick-off so it's possible they haven't looked into it too much. Alternatively, we could ask them to review the pr afterwards.